src/query/interpret/awesome_memgraph_functions.cpp

```diff
@@ -1259,6 +1263,91 @@ std::function<TypedValue(const TypedValue *, int64_t, const FunctionContext &ctx
+      mgp_memory memory{ctx.memory};
+      mgp_func_context functx{ctx.db_accessor, ctx.view};
+
+      // Check the number of arguments for a certain function
+      if (nargs < func_args.size() ||
+          // Rely on `||` short circuit so we can avoid potential overflow of
+          // func_args.size() + func_opt_args.size() by subtracting.
+          (nargs - func_args.size() > func_opt_args.size())) {
+        if (func_args.empty() && func_opt_args.empty()) {
+          throw QueryRuntimeException("'{}' requires no arguments.", fully_qualified_name);
+        } else if (func_opt_args.empty()) {
+          throw QueryRuntimeException("'{}' requires exactly {} {}.", fully_qualified_name, func_args.size(),
+                                      func_args.size() == 1U ? "argument" : "arguments");
+        } else {
+          throw QueryRuntimeException("'{}' requires between {} and {} arguments.", fully_qualified_name,
+                                      func_args.size(), func_args.size() + func_opt_args.size());
+        }
+      }
+      // The graph shouldn't be exposed to the user. Because there is no ExecutionContext,
+      // we are unable to use graph writing
+      auto graph = mgp_graph::NonWritableGraph(*ctx.db_accessor, ctx.view);
+      utils::pmr::vector<mgp_value> elems(ctx.memory);
+
+      // Build and type check function arguments.
+      for (int64_t i = 0; i < nargs; i++) {
+        auto arg = args[i];
+
+        std::string_view name;
+        const query::procedure::CypherType *type;
+        auto n_req_args = func_args.size();
+        auto n_opt_args = func_opt_args.size();
+
+        if (n_req_args > i) {
+          name = func_args[i].first;
+          type = func_args[i].second;
+        } else {
+          MG_ASSERT(n_opt_args > i - n_req_args);
+          name = std::get<0>(func_opt_args[i - n_req_args]);
+          type = std::get<1>(func_opt_args[i - n_req_args]);
+        }
+        if (!type->SatisfiesType(arg)) {
+          throw QueryRuntimeException("'{}' argument named '{}' at position {} must be of type {}.",
+                                      fully_qualified_name, name, i, type->GetPresentableName());
+        }
+        elems.emplace_back(args[i], &graph);
+      }
+
+      // Fill missing optional arguments with their default values.
+      MG_ASSERT(elems.size() >= func_args.size());
+      size_t passed_in_opt_args = nargs - func_args.size();
+      MG_ASSERT(passed_in_opt_args <= func_opt_args.size());
+      auto func_args_size = func_opt_args.size();
+      for (auto i = passed_in_opt_args; i < func_args_size; ++i) {
+        elems.emplace_back(std::get<2>(func_opt_args[i]), &graph);
+      }
+
+      auto argslist = mgp_list(std::move(elems), ctx.memory);
+
+      auto *maybe_res = func_cb(&argslist, &functx, &memory);
+      if (maybe_res->error_msg) {
+        auto error_msg = *(maybe_res->error_msg);
+        throw QueryRuntimeException(error_msg);
+      }
```

Comment:
Setting the error message could fail, so there is a chance that the user cannot indicate there is something went wrong. I think in that case the default behavior should be making sure the function call fails. E.g. the user want to set the error, but cannot, and then we try to read the `mgp_value` from the result which might be uninitialized or `nullptr`. I don't know how to solve it yet, but if you need help, feel free to ask, I am happy to help to achieve this behavior.